### PR TITLE
Fix generating xmldoc

### PR DESCRIPTION
**Pull request by vzarytovskii**
Fixes https://github.com/dotnet/fsharp/issues/14699

It seems that at some point between 2019 and 2022 we've updated a bunch of packages, which changed the positions logic somewhere, and said logic wasn't working anymore (wasn't correctly finding the line "below" the xmldoc we were about to generate), and just skipping the codegen alltogether (pretty much a couple off-by-one issues).

Second change is that we will be taking the latest document from the solution, since we were caching the first one.

And lastly, third one is that to properly delete uncommited character, we need to return to VS a `S_OK`.

~~Second issue (the diagnostic not being produced for missing params) is a different one, I will look at it separately.~~

XML docs validation works for me when I include `<WarnOn>3390;$(WarnOn)</WarnOn>` in the project file:
image

### CODE REVIEW

**Review by KevinRansom — APPROVED** [review]

**Comment by vzarytovskii** [comment]
And just to be clear, the fix is cursed. But it's probably the fastest one.
I would love to move it to something what intellisense is using.
Also, no tests,@psfinaki if you have an idea about how to test, please do share.

File: src/Compiler/Service/ServiceXmlDocParser.fs
Comment (psfinaki) [inline src/Compiler/Service/ServiceXmlDocParser.fs:1]: Changes here are just optimizations, right?

File: vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs [hunk 1]
```diff
@@ -24,12 +24,12 @@ type internal XmlDocCommandFilter
...
-    let document =
-        // There may be multiple documents with the same file path.
-        // However, for the purpose of generating XmlDoc comments, it is ok to keep only the first document.
-        lazy(match workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath) |> Seq.toList with
-             | [] -> None
-             | documentId :: _ -> Some (workspace.CurrentSolution.GetDocument documentId))
+    let getLastDocument () =
+        // There may be multiple documents with the same file path, due to versioning .
+        // However, for the purpose of generating XmlDoc comments, it is ok to keep only the last document, since they're versioned.
+        match workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath) |> Seq.toList |> List.rev with
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:30]: Why instead not something like Seq.tryLast or Seq.tryHead?

File: vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs [hunk 2]
```diff
@@ -46,6 +46,7 @@ type internal XmlDocCommandFilter
 
     interface IOleCommandTarget with
         member _.Exec(pguidCmdGroup: byref<Guid>, nCmdID: uint32, nCmdexecopt: uint32, pvaIn: IntPtr, pvaOut: IntPtr) =
+            let mutable commitCharacter = true
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:49]:
```suggestion
            let mutable isCommitCharacter = true
```

**Comment by psfinaki** [comment]
Wait, why the fix is cursed? :D

btw some formatting error

File: vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs (hunk 2 above)
Comment (vzarytovskii) [reply vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:49]: No, it's not is, it's the flag whether should we commit character or not

File: src/Compiler/Service/ServiceXmlDocParser.fs
Comment (vzarytovskii) [reply src/Compiler/Service/ServiceXmlDocParser.fs:1]: Yes

File: vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs (hunk 2 above)
Comment (psfinaki) [reply vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:49]: Right, okay. So maybe "shouldCommitCharacter" then :)

**Comment by vzarytovskii** [comment]
> Wait, why the fix is cursed? :D
> 

Doings even more magic with offsets, and adds an obscure "commit character" logic on top.

> btw some formatting error

Yeah, I haven't formatted it.

File: vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs (hunk 1 above)
Comment (0101) [inline vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:30]: Is it guaranteed to be sorted by version here or should we sort it to be safe?

**Review by 0101 — APPROVED** [review]

File: vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs (hunk 1 above)
Comment (vzarytovskii) [reply vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:30]: Dunno, didn't want to change much.

Comment (vzarytovskii) [reply vsintegration/src/FSharp.Editor/Commands/XmlDocCommandService.fs:30]: > Is it guaranteed to be sorted by version here or should we sort it to be safe?

There's nothing in the docs about it, but we have no better way as for now (before we migrate to autocomplete/snippet-like docs generation). It works now (gives correct document), might stop working in the future thought. It fixes the issue for now.

**Review by psfinaki — APPROVED** [review]
@vzarytovskii I'll look separately if this can be covered by int tests easily.

**Pull request merged**